### PR TITLE
Shape independent gradient builder for Concat

### PR DESCRIPTION
**Pull request by ashbhandare**
Current Concat gradient builder relies on shape inferencing. This change will transform Concat to ConcatTraining, and use ConcatTraining and SplitTraining to build the gradient graph.

### CODE REVIEW

File: orttraining/orttraining/test/graph/gradient_graph_builder_test.cc [hunk 1]
```diff
@@ -167,6 +169,27 @@ TEST(GradientGraphBuilderTest, BuildGradientGraphTest) {
   }
 }
 
+TEST(GradientGraphBuilderTest, BuildConcatGradientGraphTest) {
+  const auto config = MakeBasicTrainingConfig();
+  PathString backprop_model_file;
+  ASSERT_STATUS_OK(BuildBackPropGraph(CONCAT_MODEL_PATH, config, backprop_model_file));
```
Comment (ytaous) [inline orttraining/orttraining/test/graph/gradient_graph_builder_test.cc:175]: onnxruntime_test_all.exe!gsl::not_null<onnxruntime::Graph *>::get() Line 1279 C++
  onnxruntime_test_all.exe!gsl::not_null<onnxruntime::Graph *>::operator*() Line 1287 C++
  onnxruntime_test_all.exe!onnxruntime::GraphTransformer::Recurse(onnxruntime::Node & node, bool & modified, int graph_level, const onnxruntime::logging::Logger & logger) Line 47 C++
  onnxruntime_test_all.exe!onnxruntime::RuleBasedGraphTransformer::ApplyImpl(onnxruntime::Graph & graph, bool & modified, int graph_level, const onnxruntime::logging::Logger & logger) Line 84 C++
  onnxruntime_test_all.exe!onnxruntime::GraphTransformer::Apply(onnxruntime::Graph & graph, bool & modified, const onnxruntime::logging::Logger & logger) Line 14 C++
  onnxruntime_test_all.exe!onnxruntime::GraphTransformerManager::ApplyTransformers(onnxruntime::Graph & graph, onnxruntime::TransformerLevel level, const onnxruntime::logging::Logger & logger) Line 32 C++
  onnxruntime_test_all.exe!onnxruntime::training::TrainingSession::ApplyTransformationsToMainGraph(const std::unordered_set<std::string,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::string>> & weights_to_train, bool enable_gelu_approximation) Line 495 C++
  onnxruntime_test_all.exe!onnxruntime::training::TrainingSession::ConfigureForTraining(const onnxruntime::training::TrainingSession::TrainingConfiguration & config, onnxruntime::training::TrainingSession::TrainingConfigurationResult & config_result_out) Line 218 C++
  onnxruntime_test_all.exe!onnxruntime::test::BuildBackPropGraph(const std::wstring & forward_model_file, const onnxruntime::training::TrainingSession::TrainingConfiguration & config, std::wstring & backward_model_file) Line 70 C++
  onnxruntime_test_all.exe!onnxruntime::test::GradientGraphBuilderTest_BuildConcatGradientGraphTest_Test::TestBody() Line 175 C++

Comment (ytaous) [reply orttraining/orttraining/test/graph/gradient_graph_builder_test.cc:175]: failed on windows

---
In reply to: [463407164](https://github.com/microsoft/onnxruntime/pull/4675#discussion_r463407164) [](ancestors = 463407164)

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc [hunk 2]
```diff
@@ -108,8 +109,30 @@ TEST_F(GraphTransformationTests, NonZeroShapeSetter) {
   ASSERT_TRUE(nonzero_shape->dim(1).dim_param() == "nonzero_nonzero_count");
 }
 
-// MegatronF/G is defined only for training, and in msdomain.
+// MegatronF/G and ConcatTraining is defined only for training, and in msdomain.
 #ifndef DISABLE_CONTRIB_OPS
+TEST_F(GraphTransformationTests, ConcatTrainingRewriter) {
```
Comment (ytaous) [inline orttraining/orttraining/test/optimizer/graph_transform_test.cc:114]: >ConcatTrainingRewriter [](start = 33, length = 22)

same error

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc [hunk 3]
```diff
@@ -108,8 +109,30 @@ TEST_F(GraphTransformationTests, NonZeroShapeSetter) {
...
+
+  auto ret = graph_transformation_mgr.ApplyTransformers(graph, TransformerLevel::Level1, *logger_);
+  ASSERT_TRUE(ret.IsOK());
+
+  std::map<std::string, int> op_to_count = CountOpsInGraph(graph);
+
+  ASSERT_EQ(op_to_count["Concat"], 0);
+  ASSERT_EQ(op_to_count["ConcatTraining"], 1);
+
+  Model::Save(*p_model, "concat_transformed.onnx");
```
Comment (ytaous) [inline orttraining/orttraining/test/optimizer/graph_transform_test.cc:133]: do we need this line?

File: orttraining/orttraining/test/gradient/gradient_checker.cc [hunk 4]
```diff
@@ -248,7 +248,17 @@ inline Status GradientChecker<X_T, Y_T, JAC_T>::InitOpTesterWithGraph(
 
   for (size_t data_index = 0; data_index < y_infos.size(); data_index++) {
     std::string name = "output" + std::to_string(data_index);
-    op_session.AddOutput<Y_T>(name.c_str(), y_infos[data_index].shape.GetDims(), (*y_datas)[data_index]);
+    const std::vector<Y_T>& data = (*y_datas)[data_index];
+
+    if (y_infos[data_index].data_type == DataTypeImpl::GetTensorType<int64_t>()) {
```
Comment (ytaous) [inline orttraining/orttraining/test/gradient/gradient_checker.cc:253]: just curious - is this new logic for ConcatTrainingGrad only?

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#ifdef NDEBUG  // disable for debug builds because some of these tests are slow
+// #ifdef NDEBUG  // disable for debug builds because some of these tests are slow
```
Comment (ytaous) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:4]: remove //

File: orttraining/orttraining/test/graph/optimizer_graph_builder_test.cc [hunk 6]
```diff
@@ -374,7 +374,46 @@ TEST_F(OptimizerGraphBuilderTest, Allreduce_WithGradientAccumulation_WithMixedPr
   config.loss_scale_input_name = k_loss_scaling_factor_name;
   TestAllreduceOptimizerGraphBuilder(config, graph_);
 }
+TEST_F(OptimizerGraphBuilderTest, Adasum_Horovod_NoGradientAccumulation_NoMixedPrecision) {
```
Comment (ytaous) [inline orttraining/orttraining/test/graph/optimizer_graph_builder_test.cc:377]: are these new UTs relevant to concat?

Comment (ashbhandare) [reply orttraining/orttraining/test/graph/optimizer_graph_builder_test.cc:377]: they are not, had to add to get it to build locally. removing.

File: orttraining/orttraining/test/gradient/gradient_checker.cc (hunk 4 above)
Comment (ashbhandare) [reply orttraining/orttraining/test/gradient/gradient_checker.cc:253]: Yes, as we need the second output that is not available in Concat op.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 7]
```diff
@@ -491,6 +491,48 @@ IMPLEMENT_GRADIENT_BUILDER(GetConcatGradient) {
               new_attributes)};
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetConcatTrainingGradient) {
+  auto attributes = SrcNodeAttributes();
+  ORT_ENFORCE(attributes.at("axis").has_i());
```
Comment (edgchen1) [inline orttraining/orttraining/core/graph/gradient_builder.cc:496]: >has_i [](start = 36, length = 5)

i think AttributeProto.type should be used now:
[https://github.com/onnx/onnx/blob/7baca65e03fe9ec819b397fc7a796f4f93999ae8/onnx/onnx.in.proto#L151](https://github.com/onnx/onnx/blob/7baca65e03fe9ec819b397fc7a796f4f93999ae8/onnx/onnx.in.proto#L151)

there's a helper function HasInt(constAttributeProto&):
[https://github.com/microsoft/onnxruntime/blob/1eadec0eea10ad0d0f92ea9df49d67207c8d0564/onnxruntime/core/framework/tensorprotoutils.h#L157](https://github.com/microsoft/onnxruntime/blob/1eadec0eea10ad0d0f92ea9df49d67207c8d0564/onnxruntime/core/framework/tensorprotoutils.h#L157)

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 8]
```diff
@@ -491,6 +491,48 @@ IMPLEMENT_GRADIENT_BUILDER(GetConcatGradient) {
...
+  auto axis = attributes.at("axis").i();
+
+  std::vector<int64_t> split_attribute(GetSrcNodeInputSize());
+  std::vector<ArgDef> outputs;
+  bool known_shapes = true;
+  for (int i = 0; i < GetSrcNodeInputSize(); ++i) {
+    std::vector<Dimension> data_shape;
+    if (GetShape(I(i), data_shape).IsOK()) {
+      int64_t axis_index = axis < 0 ? static_cast<int64_t>(data_shape.size()) + axis : axis;
```
Comment (edgchen1) [inline orttraining/orttraining/core/graph/gradient_builder.cc:505]: >int64_t axis_index = axis < 0 ? static_cast<int64_t>(data_shape.size()) + axis : axis; [](start = 6, length = 86)

consider using HandleNegativeAxis()
[https://github.com/microsoft/onnxruntime/blob/8cf2c1c41075e455895f49e198bb3fe8e9e8231f/onnxruntime/core/providers/common.h#L17](https://github.com/microsoft/onnxruntime/blob/8cf2c1c41075e455895f49e198bb3fe8e9e8231f/onnxruntime/core/providers/common.h#L17)

File: orttraining/orttraining/core/optimizer/concat_training_rewriter.cc [hunk 9]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
```
Comment (edgchen1) [inline orttraining/orttraining/core/optimizer/concat_training_rewriter.cc:4]: includes should be ordered per:
[https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

File: orttraining/orttraining/core/optimizer/concat_training_rewriter.cc [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+#include "core/optimizer/rewrite_rule.h"
+#include "orttraining/core/optimizer/concat_training_rewriter.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status ConcatTrainingRewriter::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  auto& concat_node = node;
```
Comment (edgchen1) [inline orttraining/orttraining/core/optimizer/concat_training_rewriter.cc:14]: >auto& concat_node = node; [](start = 2, length = 25)

how about just naming the parameter `concat_node`?

File: orttraining/orttraining/core/optimizer/concat_training_rewriter.cc [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  const auto& concat_inputs = concat_node.MutableInputDefs();
+  auto& concat_outputs = concat_node.MutableOutputDefs();
+
+  ONNX_NAMESPACE::TypeProto t;
+  t.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
+  t.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(concat_inputs.size());
+
+  NodeArg& ip_shape_op = graph.GetOrCreateNodeArg("per_input_length", &t);
```
Comment (edgchen1) [inline orttraining/orttraining/core/optimizer/concat_training_rewriter.cc:22]: >"per_input_length" [](start = 50, length = 18)

should we use `graph.GenerateNodeArgName("per_input_length")`?

File: orttraining/orttraining/core/optimizer/concat_training_rewriter.cc [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+                                             "Concat with extra output",
+                                             concat_inputs,
+                                             concat_outputs,
+                                             &concat_node.GetAttributes(),
+                                             kMSDomain);
+
+  // Assign provider to this new node. Provider should be same as the provider for old node.
+  concat_training_node.SetExecutionProviderType(node.GetExecutionProviderType());
+  graph_utils::FinalizeNodeFusion(graph, concat_training_node, concat_node);
+  rule_effect = RewriteRuleEffect::kUpdatedCurrentNode;
```
Comment (edgchen1) [inline orttraining/orttraining/core/optimizer/concat_training_rewriter.cc:37]: >kUpdatedCurrentNode [](start = 35, length = 19)

i think in this case we need RewriteRuleEffect::kRemovedCurrentNode, as FinalizeNodeFusion will remove the original one

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 13]
```diff
@@ -718,7 +715,76 @@ TEST(GradientCheckerTest, ConcatGrad) {
...
     TensorShape y_shape({2, 6});
     gradient_checker.ComputeGradientError(op_def, {x1_shape, x2_shape}, {y_shape}, &max_error,
-                                          {MakeAttribute("axis", int64_t(-1))});
+                                          {MakeAttribute("axis", int64_t(-1))}, true, false);
+    EXPECT_IS_TINY(max_error);
+  }
+}
+
+TEST(GradientCheckerTest, ConcatTrainingGrad) { /*also test w/o shape inferencing */
+  float max_error;
```
Comment (edgchen1) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:724]: this code looks very similar to the previous test. can it be refactored into a test helper function?

File: orttraining/orttraining/core/optimizer/concat_training_rewriter.h [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+@Class ConcatTrainingRewriter
+
+Rewrite rule that replaces Concat with ConcatTraining, that has an additional output 
+used in building the gradient for Concat node.
+
+It is attempted to be triggered only on nodes with op type "Concat".
+*/
+class ConcatTrainingRewriter : public RewriteRule {
+ public:
+  ConcatTrainingRewriter() noexcept : RewriteRule("ConcatTrainingRewriter") {}
```
Comment (edgchen1) [inline orttraining/orttraining/core/optimizer/concat_training_rewriter.h:20]: >ConcatTrainingRewriter [](start = 51, length = 22)

"Rewrite" is somewhat generic. how about `ConcatTrainingReplacement`?

File: orttraining/orttraining/core/optimizer/concat_training_rewriter.cc (hunk 12 above)
Comment (ytaous) [reply orttraining/orttraining/core/optimizer/concat_training_rewriter.cc:37]: I tried this option, and it fixed the failed test  :-)

---
In reply to: [463870876](https://github.com/microsoft/onnxruntime/pull/4675#discussion_r463870876) [](ancestors = 463870876)

Comment (ashbhandare) [reply orttraining/orttraining/core/optimizer/concat_training_rewriter.cc:37]: Thanks!

File: orttraining/orttraining/core/optimizer/concat_training_rewriter.h (hunk 14 above)
Comment (ashbhandare) [reply orttraining/orttraining/core/optimizer/concat_training_rewriter.h:20]: Changing to "ConcatReplacement" since "ConcatTrainingReplacement" is somewhat misleading

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 15]
```diff
@@ -502,8 +504,9 @@ IMPLEMENT_GRADIENT_BUILDER(GetConcatTrainingGradient) {
   for (int i = 0; i < GetSrcNodeInputSize(); ++i) {
     std::vector<Dimension> data_shape;
     if (GetShape(I(i), data_shape).IsOK()) {
-      int64_t axis_index = axis < 0 ? static_cast<int64_t>(data_shape.size()) + axis : axis;
-      if (axis_index >= 0 && axis_index < static_cast<int64_t>(data_shape.size()) && data_shape[axis_index].has_dim_value()) {
+      int64_t rank = static_cast<int64_t>(data_shape.size());
+      int64_t axis_index = HandleNegativeAxis(axis, rank);
+      if (axis_index >= 0 && axis_index < rank && data_shape[axis_index].has_dim_value()) {
```
Comment (edgchen1) [inline orttraining/orttraining/core/graph/gradient_builder.cc:509]: `axis_index >= 0 && axis_index < rank` should already be ensured by HandleNegativeAxis()

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 16]
```diff
@@ -667,98 +667,54 @@ TEST(GradientCheckerTest, ConvGrad) {
   }
 }
 
-TEST(GradientCheckerTest, ConcatGrad) {
+void TestConcatOpGrad(const std::string& op_type,
```
Comment (edgchen1) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:670]: nit: internal linkage, e.g., `static void TestConcatOpGrad()` or `namespace { void TestConcatOpGrad(...){...} }`

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 17]
```diff
@@ -667,98 +667,54 @@ TEST(GradientCheckerTest, ConvGrad) {
...
-                                          {MakeAttribute("axis", int64_t(-1))}, true, false);
-    EXPECT_IS_TINY(max_error);
-  }
-}
-
-TEST(GradientCheckerTest, ConcatTrainingGrad) { /*also test w/o shape inferencing */
-  float max_error;
-  GradientChecker<float, float, float> gradient_checker;
-  OpDef op_def{"ConcatTraining", kMSDomain, 1};
+  bool extra_input = op_type == "ConcatTraining" ? true : false;
```
Comment (edgchen1) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:676]: consider `const bool extra_input = op_type == "ConcatTraining";`

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 18]
```diff
@@ -667,98 +667,54 @@ TEST(GradientCheckerTest, ConvGrad) {
...
 
   //concat_3d
   {
     TensorShape x_shape({1, 2, 3});
     TensorShape y_shape({1, 2, 9});
-    TensorShape y1_shape({3});
-    TensorInfo y1 = TensorInfo(y1_shape, false, nullptr, DataTypeImpl::GetTensorType<int64_t>());
+    TensorInfo y1 = TensorInfo({3}, false, nullptr, DataTypeImpl::GetTensorType<int64_t>());
+    std::vector<TensorInfo> output = {y_shape};
+    if(extra_input) output.push_back(y1);
```
Comment (edgchen1) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:713]: scope of y1 could be limited to the conditional block. nit: whitespace
this comment applies to a few other places too

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 19]
```diff
@@ -11,6 +11,8 @@
 #include "orttraining/core/framework/distributed_run_context.h"
 #include "orttraining/core/graph/gradient_builder_registry.h"
 #include "orttraining/core/graph/graph_augmenter.h"
+#include "core/providers/common.h"
```
Comment (edgchen1) [inline orttraining/orttraining/core/graph/gradient_builder.cc:14]: nit: alphabetical ordering among other ORT includes

File: onnxruntime/test/testdata/transform/concat_graph_gen.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
+import onnx
+from onnx import helper
+from onnx import TensorProto
+from enum import Enum
```
Comment (edgchen1) [inline onnxruntime/test/testdata/transform/concat_graph_gen.py:4]: unused import?
